### PR TITLE
Drag n drop sorting for custom fields

### PR DESCRIPTION
**Pull request by Kovah**
This pull request implements the ability to sort custom fields via drag n drop while editing them. The order is preserved when saving the entry.
The drag n drop feature uses the Angular CDK DragDropModule to provide frictionless support.

Preview:
bitwarden_sorting_preview

I had to move the styles for the `.box-content-row` class out of the `.box` declaration to prevent hundreds of lines code duplication. I hope this won't be a problem but I couldn't find any conflicting styles.

:warning: This pull request has a dependent pull request in the bitwarden/jslib repository which needs to be merged first before this pull request can be merged. Once it's merged I will update the submodule reference.

### CODE REVIEW

**Comment by kspearrin** [comment]
I am still confused about why there are so many style changes.

File: src/app/vault/add-edit.component.html [hunk 1]
```diff
@@ -261,36 +261,42 @@
...
-                                        [ngClass]="{'fa-eye': !f.showValue, 'fa-eye-slash': f.showValue}"></i>
+                    <div cdkDropList (cdkDropListDropped)="drop($event)">
+                        <ng-container *ngIf="cipher.hasFields">
+                            <div class="box-content-row box-content-row-multi box-draggable-row" cdkDrag
+                                *ngFor="let f of cipher.fields; let i = index; trackBy:trackByFunction"
+                                [ngClass]="{'box-content-row-checkbox': f.type === fieldType.Boolean}">
+                                <span class="box-drag-handle" title="{{'drag' | i18n}}" cdkDragHandle>
+                                    <i class="fa fa-bars"></i>
+                                </span>
+                                <a href="#" class="box-content-delete" appStopClick (click)="removeField(f)"
```
Comment (kspearrin) [inline src/app/vault/add-edit.component.html]: i dont see this `box-content-delete` class being used anywhere. why was it added?

File: src/app/vault/add-edit.component.html [hunk 2]
```diff
@@ -261,36 +261,42 @@
...
-                            </div>
-                            <input id="fieldValue{{i}}" name="Field.Value{{i}}" type="checkbox" [(ngModel)]="f.value"
-                                *ngIf="f.type === fieldType.Boolean" appTrueFalseValue trueValue="true"
-                                falseValue="false">
-                            <div class="action-buttons" *ngIf="f.type === fieldType.Hidden">
-                                <a class="row-btn" href="#" appStopClick appBlurClick
-                                    title="{{'toggleVisibility' | i18n}}" (click)="toggleFieldValue(f)">
-                                    <i class="fa fa-lg"
-                                        [ngClass]="{'fa-eye': !f.showValue, 'fa-eye-slash': f.showValue}"></i>
+                    <div cdkDropList (cdkDropListDropped)="drop($event)">
```
Comment (kspearrin) [inline src/app/vault/add-edit.component.html]: Can you explain what changed in this block? Having a hard time seeing the diff here. Looks like a new wrapping div and the `box-drag-handle` span. Is that all?

File: src/scss/box.scss [hunk 3]
```diff
@@ -1,6 +1,7 @@
 @import "variables.scss";
```
Comment (kspearrin) [inline src/scss/box.scss]: confused about why so much change

File: src/app/vault/add-edit.component.html [hunk 4]
```diff
@@ -261,36 +261,42 @@
...
+                        <ng-container *ngIf="cipher.hasFields">
+                            <div class="box-content-row box-content-row-multi box-draggable-row" cdkDrag
+                                *ngFor="let f of cipher.fields; let i = index; trackBy:trackByFunction"
+                                [ngClass]="{'box-content-row-checkbox': f.type === fieldType.Boolean}">
+                                <span class="box-drag-handle" title="{{'drag' | i18n}}" cdkDragHandle>
```
Comment (kspearrin) [inline src/app/vault/add-edit.component.html]: does a locale entry for `drag` exist?

**Comment by Kovah** [comment]
Here's an overview on the style and HTML changes:

* Wrapped the `ng-container` with a div that handles the cdkDropList, to keep the "Add Field" box out of the sorting list.
* Added a span that works as the drag handle. (Will remove the box-content-delete, used it for styling but works without it)
* Moved the whole styles for `.box-content-row` out of the `.box` block (maybe the reason for all those changes in the styles file), but there are no style changes for the box-content-row itself.
* Added a new `.box-drag-handle` block for styling the drag handle.
* Added a new `.cdk-drag-preview` block for styling the moving custom field box.

A locale entry for `drag` does not exist yet, will add it in a minute.

**Comment by kspearrin** [comment]
Thanks for the details. But, I still don't understand why `.box-content-row` needs to be outside of `.box` now. Is it because the element shown when dragging it around is outside of that scope?

**Comment by Kovah** [comment]
Correct. The dragged field is copied/moved to the end of the </body> tag and thus loses all styling of the surrounding .box element. Moving the styles out of the scope solves this and prevents duplicate styles.

**Comment by kspearrin** [comment]
Ok. I'm not a big fan of the placement of the drag handles in the current screenshot since the minus icons are now misaligned with the "add new field" plus. I might play around with moving it so that the icons can still line up.

Are you able to add this to the browser and web projects as well?

**Comment by Kovah** [comment]
Sure, no problem. Handles might be placed on the right side or wherever it makes sense.

> Are you able to add this to the browser and web projects as well?

Sure, will take a look at it as soon as possible.

**Comment by Kovah** [comment]
Just played with the placement of the handle, looks best the far right side of the field box, in my opinion. Would use this for all implementations if you are okay with it.

Screenshot 2019-03-27 at 15 22 24

**Comment by kspearrin** [comment]
yea i think i like that better

File: package-lock.json [hunk 5]
```diff
@@ -1397,12 +1406,20 @@
       "dev": true
     },
     "async": {
-      "version": "2.6.1",
-      "resolved": "https://registry.npmjs.org/async/-/async-2.6.1.tgz",
-      "integrity": "sha512-fNEiL2+AZt6AlAw/29Cr0UDe4sRAHCpEHh54WMz+Bb7QfNcFw4h3loofyJpLeQs4Yx7yuqu/2dLgM5hKOs6HlQ==",
+      "version": "2.6.2",
```
Comment (kspearrin) [inline package-lock.json]: Did you run some other npm updates here? Not sure why this package was bumped since it doesn't seem to be a dependency of cdk.

Comment (Kovah) [reply package-lock.json]: NPM seems to update packages on its own despite running npm install only. Should I revert this?

Comment (kspearrin) [reply package-lock.json]: Yes, please only make changes that are necessary for this PR.

**Pull request merged**